docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,325 @@
+# Formatted ORAS CLI output
+
+## Table of Contents
+
+- [Background](#background)
+- [Guidance](#guidance)
+- [Scenarios](#scenarios)
+  - [Scripting](#scripting)
+  - [CI/CD](#cicd)
+- [Proposal and desired user experience](#proposal-and-desired-user-experience)
+  - [oras pull](#oras-pull)
+  - [oras push](#oras-push)
+  - [oras attach](#oras-attach)
+  - [oras discover](#oras-discover)
+- [FAQ](#faq)
+
+## Background
+
+ORAS has prettified output designed for humans. However, for machine processing, especially in automation scenarios, like scripting and CI/CD pipelines, developers want to perform batch operations and chain different commands with ORAS, as well as filtering, modifying, and sorting objects based on the outputs that are emitted by the ORAS command. Developers expect that ORAS output can be emitted as machine-readable text instead of only the prettified or tabular data, so that it can be used to perform other operations.
+
+The formatted output is not intended to supersede the prettified human-readable and friendly output text of ORAS CLI. It aims to provide a programming-friendly experience for developers who want to automate their workflows without needing to parse unstructured text.
+
+## Guidance
+
+Provide two major options to enable users to define the output format of ORAS commands:
+
+- Use `--output` to output a file or directory in the filesystem
+- Use `--format` to format the output of ORAS commands into different data formats or enable process output data using the given Go template
+  - Use `--format json|tree|table` to print the output in prettified JSON, tree view, or table view
+  - Use `--format '{{ GO_TEMPLATE_FUNCTION }}'` to enable extract or compose the output data using Go template functions 
```

Comment:
My assumption is that `--format` specifies the user-friendly format like `JSON`, `tree`, `table`, 'YAML' (in the future) and also filter the data that is presented in those "formats". Current proposal assumes that I can do that in Go template but this is a steep learning curve for users that are not familiar with Go. Also overloading the `--format` parameter with two different types of inputs for the switch is confusing.

Reply:
Correct. Go template is powerful and popular but it has a bit learning curve. However, the basic filter usage with Go template is not that complicated. I listed some example usages in the [following content](https://github.com/oras-project/oras/pull/1199/files#diff-ddd85908fb8fee8c57a3726cb042b0572e9b4b333000356816268399ced78cccR109). 

There are a lot of popular CLI tools in container world like [kubectl](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#get), [docker](https://docs.docker.com/reference/cli/docker/image/ls/#options), [podman](https://docs.podman.io/en/latest/markdown/podman-images.1.html#format-format) has the same design, they also put the Go template option together with other format options in corresponding formatted output flag. I think users might be customed with this design.